### PR TITLE
adding literalDataDomains handling and other I/O field extra metadata 

### LINKED ISSUE #41: AllowedValues missing from process descriptions
The literal inputs' structure is not the same in the database as the one used in testbed14.

See: 
 - `weaver.wps_restapi.swagger_definitions.LiteralInputType`: testbed14
 - `weaver.wps_restapi.swagger_definitions.ProcessInputDescriptionSchema`: the one in the database (it's missing the allowedValues parameter, but adding it here will not add it in the process description)

So it basically comes down to 3 options I think:
 - Ignore the inputs description in the testbed14's format, and use ours instead
 - Convert our database schema to conform to testbed14's schema
 - Write a json function to support both schemas when sending the process description

I'm not sure how much work would be involved in any of these cases...

### LINKED ISSUE #211: [Bug] Missing process I/O field details
**Missing elements to fix**

- When using `/processes/...` 
  - [x] Many of the input/output fields are not returned (`dataType`, `anyValue`, `allowedValues`, etc.) (relates to #41)
  - [x] Other fields dependent on the literal data type I/O are also missing (`literalDataDomain`, types of range, etc.)
  - [x] Complex data type must be updated by modifying `mimeType` to `mediaType`. 
    Old `mimeType` format should be allowed for deployment backward-compatibility, but descriptions will be reported with the standard `mediaType` 
    (https://github.com/opengeospatial/ogcapi-processes/commit/748bba6e4d4883f116e290141b4cda26aab3bda8)

- When using `/providers/...` 
  - [x] fields available as passed literally during `DescribeProcess`, but still returns partial sub-fields any extra details such as ranges are not parsed.
  - [x] `GetCabilities` should not have the empty lists of `inputs` and `outputs` (#203) 

- [ ] Bbox data type still missing completely (#51)

References: 
https://github.com/opengeospatial/ogcapi-processes/blob/master/core/openapi/schemas
https://github.com/opengeospatial/ogcapi-processes/blob/master/core/openapi/schemas/inputDescription.yaml
https://github.com/opengeospatial/ogcapi-processes/blob/master/core/openapi/schemas/outputDescription.yaml

### LINKED ISSUE #297: Resolve invalid conversion of literal I/O extra parameters from remote WPS
## Describe the bug

Conversion of remote WPS I/O from `OWSLib` objects to `json` using `ows2json_io` function fails when it encounters `default` field of a `Literal` type. It tries to parse it as a `Complex` as if `default` was referring to the default file format. 

Error occurs here: 
https://github.com/crim-ca/weaver/blob/master/weaver/processes/convert.py#L217-L237

## To Reproduce

Sample process: 
```
https://132.217.140.46/twitcher/ows/proxy/finch?request=describeprocess&service=wps&identifier=ensemble_grid_point_cold_spell_duration_index&version=1.0.0
```
Once registered, it can be fetched in OGC-API using (adjust paths according to POSTed service registration params): 
```
http://localhost:4002/providers/finch-46/processes/ensemble_grid_point_cold_spell_duration_index
```

<details>
  <summary>
	Expand to view DescribeProcess result>
  </summary>
  <pre><code>
&lt?xml version="1.0" encoding="UTF-8"?&gt
&lt!-- PyWPS 4.4.0 --&gt
&ltwps:ProcessDescriptions xmlns:wps="http://www.opengis.net/wps/1.0.0" xmlns:ows="http://www.opengis.net/ows/1.1" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.opengis.net/wps/1.0.0 ../wpsDescribeProcess_response.xsd" service="WPS" version="1.0.0" xml:lang="en-US"&gt
    &ltProcessDescription wps:processVersion="0.1" storeSupported="true" statusSupported="true"&gt
        &ltows:Identifier&gtensemble_grid_point_cold_spell_duration_index&lt/ows:Identifier&gt
        &ltows:Title&gtCold spell duration index.&lt/ows:Title&gt
        &ltows:Abstract&gtNumber of days with at least six consecutive days where the daily minimum temperature is below the 10th percentile.&lt/ows:Abstract&gt
        &ltDataInputs&gt
            &ltInput minOccurs="1" maxOccurs="100"&gt
                &ltows:Identifier&gtlat&lt/ows:Identifier&gt
                &ltows:Title&gtLatitude&lt/ows:Title&gt
                &ltows:Abstract&gtLatitude coordinate. Accepts a comma separated list of floats for multiple grid cells.&lt/ows:Abstract&gt
                &ltLiteralData&gt
                &ltows:DataType ows:reference="http://www.w3.org/TR/xmlschema-2/#string"&gtstring&lt/ows:DataType&gt
                &lt/LiteralData&gt
            &lt/Input&gt
            &ltInput minOccurs="1" maxOccurs="100"&gt
                &ltows:Identifier&gtlon&lt/ows:Identifier&gt
                &ltows:Title&gtLongitude&lt/ows:Title&gt
                &ltows:Abstract&gtLongitude coordinate. Accepts a comma separated list of floats for multiple grid cells.&lt/ows:Abstract&gt
                &ltLiteralData&gt
                &ltows:DataType ows:reference="http://www.w3.org/TR/xmlschema-2/#string"&gtstring&lt/ows:DataType&gt
                &lt/LiteralData&gt
            &lt/Input&gt
            &ltInput minOccurs="0" maxOccurs="1"&gt
                &ltows:Identifier&gtstart_date&lt/ows:Identifier&gt
                &ltows:Title&gtInitial date&lt/ows:Title&gt
                &ltows:Abstract&gtInitial date for temporal subsetting. Can be expressed as year (%Y), year-month (%Y-%m) or year-month-day(%Y-%m-%d). Defaults to first day in file.&lt/ows:Abstract&gt
                &ltLiteralData&gt
                &ltows:DataType ows:reference="http://www.w3.org/TR/xmlschema-2/#string"&gtstring&lt/ows:DataType&gt
                &lt/LiteralData&gt
            &lt/Input&gt
            &ltInput minOccurs="0" maxOccurs="1"&gt
                &ltows:Identifier&gtend_date&lt/ows:Identifier&gt
                &ltows:Title&gtFinal date&lt/ows:Title&gt
                &ltows:Abstract&gtFinal date for temporal subsetting. Can be expressed as year (%Y), year-month (%Y-%m) or year-month-day(%Y-%m-%d). Defaults to last day in file.&lt/ows:Abstract&gt
                &ltLiteralData&gt
                &ltows:DataType ows:reference="http://www.w3.org/TR/xmlschema-2/#string"&gtstring&lt/ows:DataType&gt
                &lt/LiteralData&gt
            &lt/Input&gt
            &ltInput minOccurs="0" maxOccurs="1"&gt
                &ltows:Identifier&gtensemble_percentiles&lt/ows:Identifier&gt
                &ltows:Title&gtEnsemble percentiles&lt/ows:Title&gt
                &ltows:Abstract&gtEnsemble percentiles to calculate for input climate simulations. Accepts a comma separated list of integers.&lt/ows:Abstract&gt
                &ltLiteralData&gt
                &ltows:DataType ows:reference="http://www.w3.org/TR/xmlschema-2/#string"&gtstring&lt/ows:DataType&gt
                &ltDefaultValue&gt10,50,90&lt/DefaultValue&gt
                &lt/LiteralData&gt
            &lt/Input&gt
            &ltInput minOccurs="0" maxOccurs="1"&gt
                &ltows:Identifier&gtdataset_name&lt/ows:Identifier&gt
                &ltows:Title&gtDataset name&lt/ows:Title&gt
                &ltows:Abstract&gtName of the dataset from which to get netcdf files for inputs.&lt/ows:Abstract&gt
                &ltLiteralData&gt
                &ltows:DataType ows:reference="http://www.w3.org/TR/xmlschema-2/#string"&gtstring&lt/ows:DataType&gt
                &ltows:AllowedValues&gt
                    &ltows:Value&gtbccaqv2&lt/ows:Value&gt
                &lt/ows:AllowedValues&gt
                &lt/LiteralData&gt
            &lt/Input&gt
            &ltInput minOccurs="1" maxOccurs="1"&gt
                &ltows:Identifier&gtrcp&lt/ows:Identifier&gt
                &ltows:Title&gtRCP Scenario&lt/ows:Title&gt
                &ltows:Abstract&gtRepresentative Concentration Pathway (RCP)&lt/ows:Abstract&gt
                &ltLiteralData&gt
                &ltows:DataType ows:reference="http://www.w3.org/TR/xmlschema-2/#string"&gtstring&lt/ows:DataType&gt
                &ltows:AllowedValues&gt
                    &ltows:Value&gtrcp26&lt/ows:Value&gt
                    &ltows:Value&gtrcp45&lt/ows:Value&gt
                    &ltows:Value&gtrcp85&lt/ows:Value&gt
                &lt/ows:AllowedValues&gt
                &lt/LiteralData&gt
            &lt/Input&gt
            &ltInput minOccurs="0" maxOccurs="1000"&gt
                &ltows:Identifier&gtmodels&lt/ows:Identifier&gt
                &ltows:Title&gtModels to include in ensemble&lt/ows:Title&gt
                &ltows:Abstract&gtWhen calculating the ensemble, include only these models. By default, all 24 models are used.&lt/ows:Abstract&gt
                &ltLiteralData&gt
                &ltows:DataType ows:reference="http://www.w3.org/TR/xmlschema-2/#string"&gtstring&lt/ows:DataType&gt
                &ltows:AllowedValues&gt
                    &ltows:Value&gt24MODELS&lt/ows:Value&gt
                    &ltows:Value&gtPCIC12&lt/ows:Value&gt
                    &ltows:Value&gtBNU-ESM&lt/ows:Value&gt
                    &ltows:Value&gtCCSM4&lt/ows:Value&gt
                    &ltows:Value&gtCESM1-CAM5&lt/ows:Value&gt
                    &ltows:Value&gtCNRM-CM5&lt/ows:Value&gt
                    &ltows:Value&gtCSIRO-Mk3-6-0&lt/ows:Value&gt
                    &ltows:Value&gtCanESM2&lt/ows:Value&gt
                    &ltows:Value&gtFGOALS-g2&lt/ows:Value&gt
                    &ltows:Value&gtGFDL-CM3&lt/ows:Value&gt
                    &ltows:Value&gtGFDL-ESM2G&lt/ows:Value&gt
                    &ltows:Value&gtGFDL-ESM2M&lt/ows:Value&gt
                    &ltows:Value&gtHadGEM2-AO&lt/ows:Value&gt
                    &ltows:Value&gtHadGEM2-ES&lt/ows:Value&gt
                    &ltows:Value&gtIPSL-CM5A-LR&lt/ows:Value&gt
                    &ltows:Value&gtIPSL-CM5A-MR&lt/ows:Value&gt
                    &ltows:Value&gtMIROC-ESM-CHEM&lt/ows:Value&gt
                    &ltows:Value&gtMIROC-ESM&lt/ows:Value&gt
                    &ltows:Value&gtMIROC5&lt/ows:Value&gt
                    &ltows:Value&gtMPI-ESM-LR&lt/ows:Value&gt
                    &ltows:Value&gtMPI-ESM-MR&lt/ows:Value&gt
                    &ltows:Value&gtMRI-CGCM3&lt/ows:Value&gt
                    &ltows:Value&gtNorESM1-M&lt/ows:Value&gt
                    &ltows:Value&gtNorESM1-ME&lt/ows:Value&gt
                    &ltows:Value&gtbcc-csm1-1-m&lt/ows:Value&gt
                    &ltows:Value&gtbcc-csm1-1&lt/ows:Value&gt
                &lt/ows:AllowedValues&gt
                &ltDefaultValue&gt24MODELS&lt/DefaultValue&gt
                &lt/LiteralData&gt
            &lt/Input&gt
            &ltInput minOccurs="0" maxOccurs="1"&gt
                &ltows:Identifier&gtwindow&lt/ows:Identifier&gt
                &ltows:Title&gtWindow&lt/ows:Title&gt
                &ltows:Abstract&gtMinimum number of days with temperature below threshold to qualify as a cold spell.&lt/ows:Abstract&gt
                &ltLiteralData&gt
                &ltows:DataType ows:reference="http://www.w3.org/TR/xmlschema-2/#integer"&gtinteger&lt/ows:DataType&gt
                &ltDefaultValue&gt6&lt/DefaultValue&gt
                &lt/LiteralData&gt
            &lt/Input&gt
            &ltInput minOccurs="0" maxOccurs="1"&gt
                &ltows:Identifier&gtfreq&lt/ows:Identifier&gt
                &ltows:Title&gtFrequency&lt/ows:Title&gt
                &ltows:Abstract&gtResampling frequency.&lt/ows:Abstract&gt
                &ltLiteralData&gt
                &ltows:DataType ows:reference="http://www.w3.org/TR/xmlschema-2/#string"&gtstring&lt/ows:DataType&gt
                &ltows:AllowedValues&gt
                    &ltows:Value&gtYS&lt/ows:Value&gt
                    &ltows:Value&gtMS&lt/ows:Value&gt
                    &ltows:Value&gtQS-DEC&lt/ows:Value&gt
                    &ltows:Value&gtAS-JUL&lt/ows:Value&gt
                &lt/ows:AllowedValues&gt
                &ltDefaultValue&gtYS&lt/DefaultValue&gt
                &lt/LiteralData&gt
            &lt/Input&gt
            &ltInput minOccurs="1" maxOccurs="1"&gt
                &ltows:Identifier&gtcheck_missing&lt/ows:Identifier&gt
                &ltows:Title&gtMissing value handling method&lt/ows:Title&gt
                &ltows:Abstract&gtMethod used to determine which aggregations should be considered missing.&lt/ows:Abstract&gt
                &ltLiteralData&gt
                &ltows:DataType ows:reference="http://www.w3.org/TR/xmlschema-2/#string"&gtstring&lt/ows:DataType&gt
                &ltows:AllowedValues&gt
                    &ltows:Value&gtany&lt/ows:Value&gt
                    &ltows:Value&gtwmo&lt/ows:Value&gt
                    &ltows:Value&gtpct&lt/ows:Value&gt
                    &ltows:Value&gtat_least_n&lt/ows:Value&gt
                    &ltows:Value&gtskip&lt/ows:Value&gt
                    &ltows:Value&gtfrom_context&lt/ows:Value&gt
                &lt/ows:AllowedValues&gt
                &ltDefaultValue&gtany&lt/DefaultValue&gt
                &lt/LiteralData&gt
            &lt/Input&gt
            &ltInput minOccurs="0" maxOccurs="1"&gt
                &ltows:Identifier&gtmissing_options&lt/ows:Identifier&gt
                &ltows:Title&gtMissing method parameters&lt/ows:Title&gt
                &ltows:Abstract&gtJSON representation of dictionary of missing method parameters.&lt/ows:Abstract&gt
                &ltComplexData maximumMegabytes="200"&gt
                    &ltDefault&gt
                        &ltFormat&gt
                            &ltMimeType&gtapplication/json&lt/MimeType&gt
                        &lt/Format&gt
                    &lt/Default&gt
                    &ltSupported&gt
                        &ltFormat&gt
                            &ltMimeType&gtapplication/json&lt/MimeType&gt
                        &lt/Format&gt
                    &lt/Supported&gt
                &lt/ComplexData&gt
            &lt/Input&gt
            &ltInput minOccurs="1" maxOccurs="1"&gt
                &ltows:Identifier&gtcf_compliance&lt/ows:Identifier&gt
                &ltows:Title&gtStrictness level for CF-compliance input checks.&lt/ows:Title&gt
                &ltows:Abstract&gtWhether to log, warn or raise when inputs have non-CF-compliant attributes.&lt/ows:Abstract&gt
                &ltLiteralData&gt
                &ltows:DataType ows:reference="http://www.w3.org/TR/xmlschema-2/#string"&gtstring&lt/ows:DataType&gt
                &ltows:AllowedValues&gt
                    &ltows:Value&gtlog&lt/ows:Value&gt
                    &ltows:Value&gtwarn&lt/ows:Value&gt
                    &ltows:Value&gtraise&lt/ows:Value&gt
                &lt/ows:AllowedValues&gt
                &ltDefaultValue&gtwarn&lt/DefaultValue&gt
                &lt/LiteralData&gt
            &lt/Input&gt
            &ltInput minOccurs="1" maxOccurs="1"&gt
                &ltows:Identifier&gtdata_validation&lt/ows:Identifier&gt
                &ltows:Title&gtStrictness level for data validation input checks.&lt/ows:Title&gt
                &ltows:Abstract&gtWhether to log, warn or raise when inputs fail data validation checks.&lt/ows:Abstract&gt
                &ltLiteralData&gt
                &ltows:DataType ows:reference="http://www.w3.org/TR/xmlschema-2/#string"&gtstring&lt/ows:DataType&gt
                &ltows:AllowedValues&gt
                    &ltows:Value&gtlog&lt/ows:Value&gt
                    &ltows:Value&gtwarn&lt/ows:Value&gt
                    &ltows:Value&gtraise&lt/ows:Value&gt
                &lt/ows:AllowedValues&gt
                &ltDefaultValue&gtraise&lt/DefaultValue&gt
                &lt/LiteralData&gt
            &lt/Input&gt
            &ltInput minOccurs="0" maxOccurs="1"&gt
                &ltows:Identifier&gtoutput_format&lt/ows:Identifier&gt
                &ltows:Title&gtOutput format choice&lt/ows:Title&gt
                &ltows:Abstract&gtChoose in which format you want to recieve the result&lt/ows:Abstract&gt
                &ltLiteralData&gt
                &ltows:DataType ows:reference="http://www.w3.org/TR/xmlschema-2/#string"&gtstring&lt/ows:DataType&gt
                &ltows:AllowedValues&gt
                    &ltows:Value&gtnetcdf&lt/ows:Value&gt
                    &ltows:Value&gtcsv&lt/ows:Value&gt
                &lt/ows:AllowedValues&gt
                &ltDefaultValue&gtnetcdf&lt/DefaultValue&gt
                &lt/LiteralData&gt
            &lt/Input&gt
        &lt/DataInputs&gt
        &ltProcessOutputs&gt
            &ltOutput&gt
                &ltows:Identifier&gtoutput&lt/ows:Identifier&gt
                &ltows:Title&gtResult&lt/ows:Title&gt
                &ltows:Abstract&gtThe format depends on the &#39;output_format&#39; input parameter.&lt/ows:Abstract&gt
                &ltComplexOutput&gt
                    &ltDefault&gt
                        &ltFormat&gt
                            &ltMimeType&gtapplication/x-netcdf&lt/MimeType&gt
                            &ltEncoding&gtbase64&lt/Encoding&gt
                        &lt/Format&gt
                    &lt/Default&gt
                    &ltSupported&gt
                        &ltFormat&gt
                            &ltMimeType&gtapplication/x-netcdf&lt/MimeType&gt
                            &ltEncoding&gtbase64&lt/Encoding&gt
                        &lt/Format&gt
                        &ltFormat&gt
                            &ltMimeType&gtapplication/zip&lt/MimeType&gt
                            &ltEncoding&gtbase64&lt/Encoding&gt
                        &lt/Format&gt
                    &lt/Supported&gt
                &lt/ComplexOutput&gt
            &lt/Output&gt
            &ltOutput&gt
                &ltows:Identifier&gtoutput_log&lt/ows:Identifier&gt
                &ltows:Title&gtLogging information&lt/ows:Title&gt
                &ltows:Abstract&gtCollected logs during process run.&lt/ows:Abstract&gt
                &ltComplexOutput&gt
                    &ltDefault&gt
                        &ltFormat&gt
                            &ltMimeType&gttext/plain&lt/MimeType&gt
                        &lt/Format&gt
                    &lt/Default&gt
                    &ltSupported&gt
                        &ltFormat&gt
                            &ltMimeType&gttext/plain&lt/MimeType&gt
                        &lt/Format&gt
                    &lt/Supported&gt
                &lt/ComplexOutput&gt
            &lt/Output&gt
        &lt/ProcessOutputs&gt
    &lt/ProcessDescription&gt
&lt/wps:ProcessDescriptions&gt
  </code></pre>
</details>

The process contains many different variations of `Literal` value inputs, some with `defaultValue` (or not), others with  `allowedValues` (or not), and with different types (int, string, comma-separated string, date string, etc.)

There is also a `maximumMegabytes="200"` for a `ComplexData` input. 

## Expected behavior

Converts correctly and returns all applicable additional parameters in the description.

## References

- relates to #41 
- relates to #211

### PR DESCRIPTION
**Pull request by fmigneault**
## Changes

- Support description of additional metadata of WPS I/O of `literal` type (`defaultValue`, `literalDataDomains`, etc.). 
- Add additional `maximumMegabyte` metadata detail to `formats` of WPS I/O of `complex` type whenever available (requires  https://github.com/geopython/OWSLib/pull/796, eventual `OWSLIB==0.26.0` release).
- Other fixes applied to resolve incorrect schema validation.

Fixes #41
Fixes #211
Fixes #297

### CODE REVIEW

File: weaver/processes/convert.py [hunk 1]
```diff
@@ -224,29 +235,62 @@ def ows2json_io(ows_io):
...
+    # add value contrains specifications if missing
+    elif io_type in WPS_LITERAL_DATA_TYPE_NAMES:
+        domains = any2json_literal_data_domains(ows_io)
+        if domains:
+            json_io["literalDataDomains"] = domains
+            # fix inconsistencies of some process descriptions
+            # WPS are allowed to report 'minOccurs=1' although 'defaultValue' can also be provided
+            # (see https://github.com/bird-house/finch/pull/199)
+            if "defaultValue" in domains[0]:
+                json_io["min_occurs"] = 0
```
Comment (dbyrns) [inline weaver/processes/convert.py:293]: Do this fix get shown or for internal handling only?
My interpretation is that a default value still yields a value and thus the minimum occurrence of 1 is still met.

**Review by dbyrns — DISMISSED** [review]
👍

File: weaver/processes/convert.py (hunk 1 above)
Comment (fmigneault) [reply weaver/processes/convert.py:293]: It is displayed in the process description to make it more obvious that this value is not mandatory and that default should be used instead. During real submit, I believe omitting it on Weaver side will omit it during the remote-WPS, which could fail. 
I have patched the relevant erroneous process though directly: https://github.com/bird-house/finch/pull/199

Comment (dbyrns) [reply weaver/processes/convert.py:293]: Ok, given that minOccurs = 1 raises an error rather than using the default value, your fix is indeed welcome

**Review by matprov — APPROVED** [review]

**Pull request merged**